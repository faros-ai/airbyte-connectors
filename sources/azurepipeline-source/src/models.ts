interface Href {
  href: string;
}

interface PipelineLink {
  self: Href;
  web: Href;
}

interface RunLink {
  self: Href;
  web: Href;
  pipeline: Href;
}

interface RunPipeline {
  url: string;
  id: number;
  revision: number;
  name: string;
  folder: string;
}

interface Run {
  _links: RunLink;
  pipeline: RunPipeline;
  state: string;
  result: string;
  createdDate: string;
  finishedDate: string;
  url: string;
  id: number;
  name: string;
}

export interface RunResponse {
  count: number;
  value: Run[];
}

export interface Pipeline {
  id: number;
  revision: number;
  name?: string;
  url?: string;
  folder?: string;
  _links: PipelineLink;
  runs: Run[];
}

export interface PipelineResponse {
  count: number;
  value: Pipeline[];
}

interface BuildLink {
  self: Href;
  web: Href;
  sourceVersionDisplayUri: Href;
  timeline: Href;
  badge: Href;
}

interface BuildPlan {
  planId: string;
}

interface BuildDefinitionProject {
  id: string;
  name: string;
  description: string;
  url: string;
  state: string;
  revision: number;
  visibility: string;
  lastUpdateTime: string;
}

interface BuildDefinition {
  drafts: any[];
  id: number;
  name: string;
  url: string;
  uri: string;
  path: string;
  type: string;
  queueStatus: string;
  revision: number;
  project: BuildDefinitionProject;
}

interface BuildQueuePool {
  id: number;
  name: string;
  isHosted: boolean;
}

interface BuildQueue {
  id: number;
  name: string;
  pool: BuildQueuePool;
}

interface BuildRequestLinkAvatar {
  avatar: Href;
}

interface BuildRequestLink {
  avatar: BuildRequestLinkAvatar;
}

interface BuildRequest {
  displayName: string;
  url: string;
  id: string;
  uniqueName: string;
  imageUrl: string;
  descriptor: string;
  _link: BuildRequestLink;
}

interface BuildLog {
  id: number;
  type: string;
  url: string;
}

interface Repository {
  id: string;
  type: string;
  name: string;
  url: string;
  clean?: string;
  checkoutSubmodules: boolean;
}

interface ArtifactResource {
  data: string;
  downloadUrl: string;
  properties: any;
  type: string;
  url: string;
  _links: any;
}

interface BuildArtifact {
  id: number;
  name: string;
  resource: ArtifactResource;
  source: string;
}

export interface BuildArtifactResponse {
  count: number;
  value: BuildArtifact[];
}

interface BuildTimelineIssue {
  type: string;
  category?: any;
  message: string;
}

interface BuildTimeline {
  previousAttempts: any[];
  id: string;
  parentId: string;
  type: string;
  name: string;
  startTime?: any;
  finishTime?: any;
  currentOperation?: any;
  percentComplete?: any;
  state: string;
  result: string;
  resultCode?: any;
  changeId: number;
  lastModified: Date;
  workerName?: any;
  queueId: number;
  order: number;
  details?: any;
  errorCount: number;
  warningCount: number;
  url?: any;
  log?: any;
  task?: any;
  attempt: number;
  identifier: string;
  issues: BuildTimelineIssue[];
}

export interface BuildTimelineResponse {
  records: BuildTimeline[];
}

export interface Build {
  _links: BuildLink;
  properties: any;
  tags: any[];
  validationResults: any[];
  plans: BuildPlan[];
  triggerInfo: any;
  id: number;
  buildNumber: string;
  status: string;
  result: string;
  queueTime: string;
  startTime: string;
  finishTime: string;
  url: string;
  definition: BuildDefinition;
  buildNumberRevision: number;
  project: BuildDefinitionProject;
  uri: string;
  sourceBranch: string;
  sourceVersion: string;
  queue: BuildQueue;
  priority: string;
  reason: string;
  requestedFor: BuildRequest;
  requestedBy: BuildRequest;
  lastChangedDate: string;
  lastChangedBy: BuildRequest;
  orchestrationPlan: BuildPlan;
  logs: BuildLog;
  repository: Repository;
  keepForever: boolean;
  retainedByRelease: boolean;
  triggeredByBuild: any;
  artifacts: BuildArtifact[];
  jobs: BuildTimeline[];
}

export interface BuildResponse {
  count: number;
  value: Build[];
}

interface ReleaseByUser {
  id: string;
  displayName: string;
  uniqueName: string;
  url: string;
  imageUrl?: string;
}

interface ReleaseLink {
  self: Href;
  web: Href;
}

interface ReleaseProjectReference {
  id: string;
  name: string;
}

export interface Release {
  id: number;
  name: string;
  status: string;
  createdOn: string;
  modifiedOn?: string;
  modifiedBy?: ReleaseByUser;
  createdBy: ReleaseByUser;
  variables: any;
  variableGroups: [any];
  description?: string;
  reason?: string;
  releaseNameFormat?: string;
  keepForever: boolean;
  definitionSnapshotRevision: number;
  logsContainerUrl: string;
  url: string;
  _links: ReleaseLink;
  tags: [any];
  projectReference: ReleaseProjectReference;
  properties: any;
}

export interface ReleaseResponse {
  count: number;
  value: Release[];
}
