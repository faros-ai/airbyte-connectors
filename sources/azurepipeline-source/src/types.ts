import {
  Build as AzureBuild,
  BuildArtifact,
} from 'azure-devops-node-api/interfaces/BuildInterfaces';
import {CodeCoverageStatistics} from 'azure-devops-node-api/interfaces/TestInterfaces';
import {
  AzureDevOpsConfig,
  TimelineRecord,
} from 'faros-airbyte-common/azure-devops';

export interface AzurePipelineConfig extends AzureDevOpsConfig {}

export interface Build extends AzureBuild {
  artifacts: BuildArtifact[];
  coverageStats: CodeCoverageStatistics[];
  stages: TimelineRecord[];
}
