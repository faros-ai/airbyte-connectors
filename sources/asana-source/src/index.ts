import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
