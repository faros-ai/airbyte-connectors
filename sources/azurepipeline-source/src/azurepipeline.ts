import {AxiosInstance} from 'axios';
import {AirbyteLogger, base64Encode, wrapApiError} from 'faros-airbyte-cdk';
import {makeAxiosInstanceWithRetry} from 'faros-js-client';
import {VError} from 'verror';

import {
  Build,
  BuildArtifactResponse,
  BuildResponse,
  BuildTimelineResponse,
  Pipeline,
  PipelineResponse,
  Release,
  ReleaseResponse,
} from './models';

const DEFAULT_API_VERSION = '6.0';
const DEFAULT_CUTOFF_DAYS = 90;
const DEFAULT_PAGE_SIZE = 100;
const DEFAULT_API_TIMEOUT_MS = 0; // 0 means no timeout
const DEFAULT_RETRIES = 3;
const DEFAULT_RETRY_DELAY_MS = 1000;
const CONTINUATION_TOKEN_HEADER = 'x-ms-continuationtoken';

export interface AzurePipelineConfig {
  readonly organization: string;
  readonly projects: string[];
  readonly access_token: string;
  readonly cutoff_days?: number;
  readonly page_size?: number;
  readonly api_version?: string;
  readonly api_timeout?: number;
  readonly max_retries?: number;
  readonly api_retry_delay?: number;
}

export class AzurePipeline {
  private static azurePipeline: AzurePipeline = null;

  constructor(
    private readonly httpClient: AxiosInstance,
    private readonly httpVSRMClient: AxiosInstance,
    private readonly startDate: Date,
    private readonly projects: ReadonlyArray<string>,
    private readonly pageSize: number
  ) {}

  static instance(
    config: AzurePipelineConfig,
    logger?: AirbyteLogger
  ): AzurePipeline {
    if (AzurePipeline.azurePipeline) return AzurePipeline.azurePipeline;

    if (!config.access_token) {
      throw new VError('Please provide an access token');
    }

    if (!config.organization) {
      throw new VError('Please provide an organization');
    }

    if (!config.projects || config.projects.length === 0) {
      throw new VError('Please provide at least one project name');
    }

    const cutoff_days = config.cutoff_days ?? DEFAULT_CUTOFF_DAYS;

    const accessToken = base64Encode(`:${config.access_token}`);

    const startDate = new Date();
    startDate.setDate(startDate.getDate() - cutoff_days);

    const version = config.api_version ?? DEFAULT_API_VERSION;

    const httpClient = makeAxiosInstanceWithRetry(
      {
        baseURL: `https://dev.azure.com/${config.organization}`,
        timeout: config.api_timeout ?? DEFAULT_API_TIMEOUT_MS,
        maxContentLength: Infinity, //default is 2000 bytes
        params: {
          'api-version': version,
        },
        headers: {
          Authorization: `Basic ${accessToken}`,
        },
      },
      logger?.asPino(),
      config.max_retries ?? DEFAULT_RETRIES,
      config.api_retry_delay ?? DEFAULT_RETRY_DELAY_MS
    );

    const httpVSRMClient = makeAxiosInstanceWithRetry(
      {
        baseURL: `https://vsrm.dev.azure.com/${config.organization}`,
        timeout: config.api_timeout ?? DEFAULT_API_TIMEOUT_MS,
        maxContentLength: Infinity, //default is 2000 bytes
        params: {
          'api-version': version,
        },
        headers: {
          Authorization: `Basic ${accessToken}`,
        },
      },
      logger?.asPino(),
      config.max_retries ?? DEFAULT_RETRIES,
      config.api_retry_delay ?? DEFAULT_RETRY_DELAY_MS
    );

    AzurePipeline.azurePipeline = new AzurePipeline(
      httpClient,
      httpVSRMClient,
      startDate,
      config.projects,
      config.page_size ?? DEFAULT_PAGE_SIZE
    );

    return AzurePipeline.azurePipeline;
  }

  async checkConnection(): Promise<void> {
    try {
      const iter = this.getPipelines(this.projects[0]);
      await iter.next();
    } catch (err: any) {
      let errorMessage = 'Please verify your access token is correct. Error: ';
      if (err.error_code || err.error_info) {
        errorMessage += `${err.error_code}: ${err.error_info}`;
        throw new VError(errorMessage);
      }
      try {
        errorMessage += err.message ?? err.statusText ?? wrapApiError(err);
      } catch (wrapError: any) {
        errorMessage += wrapError.message;
      }
      throw new VError(errorMessage);
    }
  }

  async *getPipelines(
    project: string,
    logger?: AirbyteLogger
  ): AsyncGenerator<Pipeline> {
    const params = {$top: this.pageSize};
    let hasNext = true;
    let continuationToken = undefined;

    while (hasNext) {
      const res = await this.httpClient.get<PipelineResponse>(
        `${project}/_apis/pipelines`,
        {params}
      );

      logger?.info(`Fetched ${res.data.count} pipelines`);

      for (const pipeline of res.data.value) {
        yield pipeline;
      }

      continuationToken = res.headers[CONTINUATION_TOKEN_HEADER];
      hasNext = continuationToken !== undefined && continuationToken !== '';
      logger?.info(
        hasNext ? 'Fetching next pipelines page' : 'No more pipelines'
      );
      params['continuationToken'] = continuationToken;
    }
  }

  async *getBuilds(
    project: string,
    lastQueueTime?: string,
    logger?: AirbyteLogger
  ): AsyncGenerator<Build> {
    const startTime = lastQueueTime ? new Date(lastQueueTime) : this.startDate;
    //https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/list?view=azure-devops-rest-6.0
    //https://docs.microsoft.com/en-us/rest/api/azure/devops/build/builds/list?view=azure-devops-rest-6.0#buildqueryorder
    const params = {
      queryOrder: 'queueTimeAscending',
      minTime: startTime.toISOString(),
      $top: this.pageSize,
    };
    let hasNext = true;
    let continuationToken = undefined;

    while (hasNext) {
      const res = await this.httpClient.get<BuildResponse>(
        `${project}/_apis/build/builds`,
        {params}
      );

      logger?.info(`Fetched ${res.data.count} builds`);

      for (const item of res.data.value) {
        const artifact = await this.httpClient.get<BuildArtifactResponse>(
          `${project}/_apis/build/builds/${item.id}/artifacts`
        );

        if (artifact.status === 200) {
          item.artifacts = artifact.data.value;
        }

        const timeline = await this.httpClient.get<BuildTimelineResponse>(
          `${project}/_apis/build/builds/${item.id}/timeline`
        );

        const timelines = [];
        if (timeline.status === 200) {
          for (const item of timeline.data.records) {
            if (item.type === 'Job') timelines.push(item);
          }
        }

        item.jobs = timelines;
        yield item;
      }

      continuationToken = res.headers[CONTINUATION_TOKEN_HEADER];
      hasNext = continuationToken !== undefined && continuationToken !== '';
      logger?.info(hasNext ? 'Fetching next builds page' : 'No more builds');
      params['continuationToken'] = continuationToken;
    }
  }

  async *getReleases(
    project: string,
    lastCreatedOn?: string,
    logger?: AirbyteLogger
  ): AsyncGenerator<Release> {
    const startTime = lastCreatedOn ? new Date(lastCreatedOn) : this.startDate;
    //https://docs.microsoft.com/en-us/rest/api/azure/devops/release/releases/list?view=azure-devops-rest-6.0
    //https://docs.microsoft.com/en-us/rest/api/azure/devops/release/releases/list?view=azure-devops-rest-6.0#releasequeryorder
    const params = {
      queryOrder: 'ascending',
      minCreatedTime: startTime.toISOString(),
      $top: this.pageSize,
    };
    let hasNext = true;
    let continuationToken = undefined;

    while (hasNext) {
      const res = await this.httpVSRMClient.get<ReleaseResponse>(
        `${project}/_apis/release/releases`,
        {params}
      );

      logger?.info(`Fetched ${res.data.count} releases`);

      for (const release of res.data.value) {
        yield release;
      }

      continuationToken = res.headers[CONTINUATION_TOKEN_HEADER];
      hasNext = continuationToken !== undefined && continuationToken !== '';
      logger?.info(
        hasNext ? 'Fetching next releases page' : 'No more releases'
      );
      params['continuationToken'] = continuationToken;
    }
  }
}
