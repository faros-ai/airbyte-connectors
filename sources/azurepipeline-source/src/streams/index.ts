import {Builds} from './builds';
import {Pipelines} from './pipelines';
import {Releases} from './releases';
export {Pipelines, Builds, Releases};
