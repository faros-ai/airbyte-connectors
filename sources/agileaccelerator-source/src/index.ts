import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
