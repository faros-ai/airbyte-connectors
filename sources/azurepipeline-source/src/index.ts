import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzurePipeline, AzurePipelineConfig} from './azurepipeline';
import {Builds, Pipelines, Releases} from './streams';
/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzurePipelineSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzurePipeline source implementation. */
export class AzurePipelineSource extends AirbyteSourceBase<AzurePipelineConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AzurePipelineConfig
  ): Promise<[boolean, VError]> {
    try {
      const azureActiveDirectory = AzurePipeline.instance(config);
      await azureActiveDirectory.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzurePipelineConfig): AirbyteStreamBase[] {
    return [
      new Pipelines(config, this.logger),
      new Builds(config, this.logger),
      new Releases(config, this.logger),
    ];
  }
}
