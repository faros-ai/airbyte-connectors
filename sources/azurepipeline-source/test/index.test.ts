import {
  AirbyteLogger,
  AirbyteLogLevel,
  AirbyteSpec,
  SyncMode,
} from 'faros-airbyte-cdk';
import fs from 'fs-extra';
import {VError} from 'verror';

import {AzurePipeline} from '../src/azurepipeline';
import * as sut from '../src/index';

const azureActivePipeline = AzurePipeline.instance;

describe('index', () => {
  const logger = new AirbyteLogger(
    // Shush messages in tests, unless in debug
    process.env.LOG_LEVEL === 'debug'
      ? AirbyteLogLevel.DEBUG
      : AirbyteLogLevel.FATAL
  );

  beforeEach(() => {
    AzurePipeline.instance = azureActivePipeline;
  });

  function readResourceFile(fileName: string): any {
    return JSON.parse(fs.readFileSync(`resources/${fileName}`, 'utf8'));
  }

  function readTestResourceFile(fileName: string): any {
    return JSON.parse(fs.readFileSync(`test_files/${fileName}`, 'utf8'));
  }
  test('spec', async () => {
    const source = new sut.AzurePipelineSource(logger);
    await expect(source.spec()).resolves.toStrictEqual(
      new AirbyteSpec(readResourceFile('spec.json'))
    );
  });

  test('check connection - no access token', async () => {
    const source = new sut.AzurePipelineSource(logger);
    await expect(
      source.checkConnection({
        access_token: '',
        organization: 'organization',
        project: 'project',
        cutoff_days: 365,
      } as any)
    ).resolves.toStrictEqual([
      false,
      new VError('access_token must not be an empty string'),
    ]);
  });

  test('streams - pipelines, use full_refresh sync mode', async () => {
    const projectNames = ['project'];
    const fnPipelinesFunc = jest.fn();

    AzurePipeline.instance = jest.fn().mockImplementation(() => {
      const pipelinesResource: any[] = readTestResourceFile('pipelines.json');
      return new AzurePipeline(
        {
          get: fnPipelinesFunc.mockResolvedValue({
            data: {value: pipelinesResource},
          }),
        } as any,
        null,
        new Date('2010-03-27T14:03:51-0800'),
        projectNames
      );
    });
    const source = new sut.AzurePipelineSource(logger);
    const streams = source.streams({} as any);

    const pipelinesStream = streams[0];
    const pipelineIter = pipelinesStream.readRecords(
      SyncMode.FULL_REFRESH,
      undefined,
      {project: projectNames[0]}
    );
    const pipelines = [];
    for await (const pipeline of pipelineIter) {
      pipelines.push(pipeline);
    }

    expect(fnPipelinesFunc).toHaveBeenCalledTimes(3);
    expect(pipelines).toStrictEqual(readTestResourceFile('pipelines.json'));
  });

  test('streams - builds, use full_refresh sync mode', async () => {
    const projectNames = ['project'];
    const fnBuildsFunc = jest.fn();

    AzurePipeline.instance = jest.fn().mockImplementation(() => {
      const buildsResource: any[] = readTestResourceFile('builds.json');
      return new AzurePipeline(
        {
          get: fnBuildsFunc.mockResolvedValue({
            data: {value: buildsResource},
          }),
        } as any,
        null,
        new Date('2010-03-27T14:03:51-0800'),
        projectNames
      );
    });
    const source = new sut.AzurePipelineSource(logger);
    const streams = source.streams({} as any);

    const buildsStream = streams[0];
    const buildIter = buildsStream.readRecords(
      SyncMode.FULL_REFRESH,
      undefined,
      {project: projectNames[0]}
    );
    const builds = [];
    for await (const build of buildIter) {
      builds.push(build);
    }

    expect(fnBuildsFunc).toHaveBeenCalledTimes(4);
    expect(builds).toStrictEqual(readTestResourceFile('builds.json'));
  });

  test('streams - releases, use full_refresh sync mode', async () => {
    const projectNames = ['project'];
    const fnReleasesFunc = jest.fn();

    AzurePipeline.instance = jest.fn().mockImplementation(() => {
      const releasesResource: any[] = readTestResourceFile('releases.json');
      return new AzurePipeline(
        {
          get: fnReleasesFunc.mockResolvedValue({
            data: {value: releasesResource},
          }),
        } as any,
        null,
        new Date('2010-03-27T14:03:51-0800'),
        projectNames
      );
    });
    const source = new sut.AzurePipelineSource(logger);
    const streams = source.streams({} as any);

    const releasesStream = streams[0];
    const releaseIter = releasesStream.readRecords(
      SyncMode.FULL_REFRESH,
      undefined,
      {project: projectNames[0]}
    );
    const releases = [];
    for await (const release of releaseIter) {
      releases.push(release);
    }

    expect(fnReleasesFunc).toHaveBeenCalledTimes(2);
    expect(releases).toStrictEqual(readTestResourceFile('releases.json'));
  });
});
