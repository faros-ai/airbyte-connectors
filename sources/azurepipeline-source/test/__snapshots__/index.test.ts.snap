// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index check connection - no access token 1`] = `
Array [
  false,
  [VError: Please provide an access token],
]
`;

exports[`index streams - builds with coverage 1`] = `
Array [
  Object {
    "_links": Object {
      "badge": Object {
        "href": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/status/2",
      },
      "self": Object {
        "href": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/Builds/24",
      },
      "sourceVersionDisplayUri": Object {
        "href": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/builds/24/sources",
      },
      "timeline": Object {
        "href": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/builds/24/Timeline",
      },
      "web": Object {
        "href": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_build/results?buildId=24",
      },
    },
    "artifacts": Array [],
    "buildNumber": "20220214.1",
    "buildNumberRevision": 1,
    "coverageStats": Array [
      Object {
        "covered": 446,
        "delta": 0,
        "isDeltaAvailable": false,
        "label": "Lines",
        "position": 4,
        "total": 708,
      },
    ],
    "definition": Object {
      "drafts": Array [],
      "id": 2,
      "name": "coonglinh.medee_dms",
      "path": "\\",
      "project": Object {
        "description": "tesst",
        "id": "a432baaf-0264-4311-b613-087a2e7931f8",
        "lastUpdateTime": "2022-02-10T07:00:49.76Z",
        "name": "devcube",
        "revision": 12,
        "state": "wellFormed",
        "url": "https://dev.azure.com/huongkstn/_apis/projects/a432baaf-0264-4311-b613-087a2e7931f8",
        "visibility": "private",
      },
      "queueStatus": "enabled",
      "revision": 1,
      "type": "build",
      "uri": "vstfs:///Build/Definition/2",
      "url": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/Definitions/2?revision=1",
    },
    "finishTime": "2022-02-14T02:35:38.7767368Z",
    "id": 24,
    "jobs": Array [
      Object {
        "attempt": 1,
        "changeId": 9,
        "currentOperation": null,
        "details": null,
        "errorCount": 1,
        "finishTime": null,
        "id": "b44a04b4-b5b2-595a-77f4-4561138d2268",
        "identifier": "A.__default",
        "issues": Array [
          Object {
            "category": null,
            "message": "No hosted parallelism has been purchased or granted. To request a free parallelism grant, please fill out the following form https://aka.ms/azpipelines-parallelism-request",
            "type": "error",
          },
        ],
        "lastModified": "0001-01-01T00:00:00",
        "log": null,
        "name": "A",
        "order": 1,
        "parentId": "67c760f8-35f0-533f-1d24-8e8c3788c96d",
        "percentComplete": null,
        "previousAttempts": Array [],
        "queueId": 9,
        "result": "failed",
        "resultCode": null,
        "startTime": null,
        "state": "completed",
        "task": null,
        "type": "Job",
        "url": null,
        "warningCount": 0,
        "workerName": null,
      },
      Object {
        "attempt": 1,
        "changeId": 10,
        "currentOperation": null,
        "details": null,
        "errorCount": 1,
        "finishTime": null,
        "id": "b79b690b-82d1-5750-8a0a-452d70195841",
        "identifier": "B.__default",
        "issues": Array [
          Object {
            "category": null,
            "message": "No hosted parallelism has been purchased or granted. To request a free parallelism grant, please fill out the following form https://aka.ms/azpipelines-parallelism-request",
            "type": "error",
          },
        ],
        "lastModified": "0001-01-01T00:00:00",
        "log": null,
        "name": "B",
        "order": 1,
        "parentId": "c8ede3e9-af9a-5ef1-8d3f-38453e19c638",
        "percentComplete": null,
        "previousAttempts": Array [],
        "queueId": 9,
        "result": "failed",
        "resultCode": null,
        "startTime": null,
        "state": "completed",
        "task": null,
        "type": "Job",
        "url": null,
        "warningCount": 0,
        "workerName": null,
      },
    ],
    "keepForever": false,
    "lastChangedBy": Object {
      "_links": Object {
        "avatar": Object {
          "href": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/s2s.MDAwMDAwMDItMDAwMC04ODg4LTgwMDAtMDAwMDAwMDAwMDAwQDJjODk1OTA4LTA0ZTAtNDk1Mi04OWZkLTU0YjAwNDZkNjI4OA",
        },
      },
      "descriptor": "s2s.MDAwMDAwMDItMDAwMC04ODg4LTgwMDAtMDAwMDAwMDAwMDAwQDJjODk1OTA4LTA0ZTAtNDk1Mi04OWZkLTU0YjAwNDZkNjI4OA",
      "displayName": "Microsoft.VisualStudio.Services.TFS",
      "id": "00000002-0000-8888-8000-000000000000",
      "imageUrl": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/s2s.MDAwMDAwMDItMDAwMC04ODg4LTgwMDAtMDAwMDAwMDAwMDAwQDJjODk1OTA4LTA0ZTAtNDk1Mi04OWZkLTU0YjAwNDZkNjI4OA",
      "uniqueName": "00000002-0000-8888-8000-000000000000@2c895908-04e0-4952-89fd-54b0046d6288",
      "url": "https://spsprodcus5.vssps.visualstudio.com/Adc21dd72-1c92-418a-bea3-e51c7a562f87/_apis/Identities/00000002-0000-8888-8000-000000000000",
    },
    "lastChangedDate": "2022-02-14T02:35:38.867Z",
    "logs": Object {
      "id": 0,
      "type": "Container",
      "url": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/builds/24/logs",
    },
    "orchestrationPlan": Object {
      "planId": "d5fd24b3-2524-45b8-b641-a19bee1a8308",
    },
    "plans": Array [
      Object {
        "planId": "d5fd24b3-2524-45b8-b641-a19bee1a8308",
      },
    ],
    "priority": "normal",
    "project": Object {
      "description": "tesst",
      "id": "a432baaf-0264-4311-b613-087a2e7931f8",
      "lastUpdateTime": "2022-02-10T07:00:49.76Z",
      "name": "devcube",
      "revision": 12,
      "state": "wellFormed",
      "url": "https://dev.azure.com/huongkstn/_apis/projects/a432baaf-0264-4311-b613-087a2e7931f8",
      "visibility": "private",
    },
    "properties": Object {},
    "queue": Object {
      "id": 9,
      "name": "Azure Pipelines",
      "pool": Object {
        "id": 9,
        "isHosted": true,
        "name": "Azure Pipelines",
      },
    },
    "queueTime": "2022-02-14T02:35:32.5194025Z",
    "reason": "pullRequest",
    "repository": Object {
      "checkoutSubmodules": false,
      "clean": null,
      "id": "coonglinh/medee_dms",
      "type": "Bitbucket",
    },
    "requestedBy": Object {
      "_links": Object {
        "avatar": Object {
          "href": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
        },
      },
      "descriptor": "msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
      "displayName": "tran ngoc huong",
      "id": "8725d325-8926-6514-a563-8717451ce799",
      "imageUrl": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
      "uniqueName": "huongkstn@gmail.com",
      "url": "https://spsprodcus5.vssps.visualstudio.com/Adc21dd72-1c92-418a-bea3-e51c7a562f87/_apis/Identities/8725d325-8926-6514-a563-8717451ce799",
    },
    "requestedFor": Object {
      "_links": Object {
        "avatar": Object {
          "href": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
        },
      },
      "descriptor": "msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
      "displayName": "tran ngoc huong",
      "id": "8725d325-8926-6514-a563-8717451ce799",
      "imageUrl": "https://dev.azure.com/huongkstn/_apis/GraphProfile/MemberAvatars/msa.ODcyNWQzMjUtODkyNi03NTE0LWE1NjMtODcxNzQ1MWNlNzk5",
      "uniqueName": "huongkstn@gmail.com",
      "url": "https://spsprodcus5.vssps.visualstudio.com/Adc21dd72-1c92-418a-bea3-e51c7a562f87/_apis/Identities/8725d325-8926-6514-a563-8717451ce799",
    },
    "result": "succeeded",
    "retainedByRelease": false,
    "sourceBranch": "refs/heads/master",
    "sourceVersion": "d6ef20d64886947134ea0bcdcadce24f8b7823ff",
    "startTime": "2022-02-14T02:35:38.7767368Z",
    "status": "completed",
    "tags": Array [],
    "triggerInfo": Object {},
    "triggeredByBuild": null,
    "uri": "vstfs:///Build/Build/24",
    "url": "https://dev.azure.com/huongkstn/a432baaf-0264-4311-b613-087a2e7931f8/_apis/build/Builds/24",
    "validationResults": Array [],
  },
]
`;
